fedimint-core/src/encoding/mod.rs

```diff
@@ -1004,11 +1004,70 @@ impl Decodable for Cow<'static, str> {
 // Simple decoder implementing `bitcoin_io::Read` for `std::io::Read`.
 // This is needed because `bitcoin::consensus::Decodable` requires a
 // `bitcoin_io::Read`.
-pub struct SimpleBitcoinRead<R: std::io::Read>(R);
+struct SimpleBitcoinRead<'a, R: Read> {
+    inner: &'a mut R,
+    buf: [u8; 1],
+    is_consumed: bool,
+}
+
+impl<'a, R: Read> SimpleBitcoinRead<'a, R> {
+    /// Creates a [`SimpleBitcoinRead`] which will read from the given `inner`.
+    pub fn new(inner: &'a mut R) -> Self {
+        SimpleBitcoinRead {
+            inner,
+            buf: [0; 1],
+            is_consumed: true,
+        }
+    }
+}
+
+impl<'a, R: Read> bitcoin_io::Read for SimpleBitcoinRead<'a, R> {
+    #[inline]
+    fn read(&mut self, output: &mut [u8]) -> bitcoin_io::Result<usize> {
+        if output.is_empty() {
+            return Ok(0);
+        }
+        #[allow(clippy::useless_let_if_seq)]
+        let mut offset = 0;
+        if !self.is_consumed {
+            output[0] = self.buf[0];
+            self.is_consumed = true;
+            offset = 1;
+        }
+        Ok(self
+            .inner
+            .read(&mut output[offset..])
+            .map(|len| len + offset)?)
+    }
+}
 
-impl<R: std::io::Read> bitcoin_io::Read for SimpleBitcoinRead<R> {
-    fn read(&mut self, buf: &mut [u8]) -> bitcoin_io::Result<usize> {
-        self.0.read(buf).map_err(bitcoin_io::Error::from)
```

Comment:
I wouldn't remove this impl, it's much better. The fake `BufRead` impl should only be used where absolutely necessary.

Reply:
Wouldn't that make it unsafe to switch between using `bitcoin_io::Read` and `bitcoin_io::BufRead`? I know we're not actually doing that, but if we ever needed to, that could lead to very wonky behavior.

Reply:
Why would it be unsafe? You can use `SimpleBitcoinRead` or `FakeBufRead` where `bitcoin_io::Read` is required and use `FakeBufRead` where `bitcoin_io::BufRead` is required. I'd just prefer to use `SimpleBitcoinRead` wherever possible since it allows reading multiple bytes at once if one knows how many bytes are expected. `FakeBufRead` always has to go byte by byte.

Reply:
Discussed offline. Eric was thinking we add a separate implementation for `BufRead` whereas I thought we were extending `SimpleBitcoinRead`, in which case we would need the more complex implementation. We're going with the former.